util/fee-estimator/src/estimator/weight_units_flow.rs

```diff
@@ -0,0 +1,463 @@
+        }
+        let item = TxStatus::new_from_entry_info(info);
+        self.txs
+            .entry(self.current_tip)
+            .and_modify(|items| items.push(item))
+            .or_insert_with(|| vec![item]);
+    }
+
+    pub fn get_fee_estimates(
+        &self,
+        all_entry_info: TxPoolEntryInfo,
+    ) -> Result<Option<RecommendedFeeRates>, Error> {
+        if !self.is_ready {
+            return Err(Error::NotReady);
+        }
+
+        let sorted_current_txs = {
+            let mut current_txs: Vec<_> = all_entry_info
+                .pending
+                .into_values()
+                .chain(all_entry_info.proposed.into_values())
+                .map(TxStatus::new_from_entry_info)
+                .collect();
+            current_txs.sort_unstable_by(|a, b| b.cmp(a));
+            current_txs
+        };
+
+        let high = if let Some(fee_rate) =
+            self.do_estimate(constants::HIGH_TARGET, &sorted_current_txs)?
+        {
+            fee_rate
+        } else {
+            return Ok(None);
+        };
+
+        let medium = if let Ok(Some(fee_rate)) =
+            self.do_estimate(constants::MEDIUM_TARGET, &sorted_current_txs)
+        {
+            fee_rate
+        } else {
+            let rates = RecommendedFeeRates {
+                default: high,
+                low: high,
+                medium: high,
+                high,
+            };
+            return Ok(Some(rates));
+        };
+
+        let low = if let Ok(Some(fee_rate)) =
+            self.do_estimate(constants::LOW_TARGET, &sorted_current_txs)
+        {
+            fee_rate
+        } else {
+            let rates = RecommendedFeeRates {
+                default: medium,
+                low: medium,
+                medium,
+                high,
+            };
+            return Ok(Some(rates));
+        };
+
+        let default = if let Ok(Some(fee_rate)) =
+            self.do_estimate(constants::DEFAULT_TARGET, &sorted_current_txs)
+        {
+            fee_rate
+        } else {
+            let rates = RecommendedFeeRates {
+                default: low,
+                low,
+                medium,
+                high,
+            };
+            return Ok(Some(rates));
+        };
+
+        let rates = RecommendedFeeRates {
+            default,
+            low,
+            medium,
+            high,
+        };
+        Ok(Some(rates))
+    }
+}
+
+impl Algorithm {
+    fn do_estimate(
+        &self,
+        target_blocks: BlockNumber,
+        sorted_current_txs: &[TxStatus],
+    ) -> Result<Option<FeeRate>, Error> {
+        ckb_logger::debug!(
+            "boot: {}, current: {}, target: {target_blocks} blocks",
+            self.boot_tip,
+            self.current_tip,
+        );
+        let historical_blocks = Self::historical_blocks(target_blocks);
+        ckb_logger::debug!("required: {historical_blocks} blocks");
+        if historical_blocks > self.current_tip - self.boot_tip {
```

Comment:
Is it possible for current_tip to be lower than boot_tip? How about using `saturating_sub` here?